src/qrules/io/_dot.py

```diff
@@ -497,26 +494,39 @@ def _collapse_graphs(
         )
         for g in graphs
     }
+
     for transition in graphs:
         topology = transition.topology
         group = transition_groups[topology]
         for state_id, state in transition.states.items():
-            if isinstance(state, State):
-                particle = state.particle
-            else:
-                particle, _ = state
-            group.states[state_id].add(particle)
+            group.states[state_id].add(_strip_properties(state))
     collected_graphs: list[FrozenTransition[tuple[Particle, ...], None]] = []
     for topology in sorted(transition_groups):
         group = transition_groups[topology]
         collected_graphs.append(
             FrozenTransition(
                 topology,
                 states={
-                    i: tuple(sorted(particles, key=lambda p: p.name))
+                    i: tuple(sorted(particles, key=_gen_sorting_key(group.states[i])))
                     for i, particles in group.states.items()
                 },
                 interactions=group.interactions,
             )
         )
     return collected_graphs
+
+
+def _strip_properties(state: Any) -> Any:
+    if isinstance(state, State):
+        return state.particle
+    if isinstance(state, str):
+        return state
+    return state
+
+
+def _gen_sorting_key(obj: Any) -> Callable[[Any], str] | None:
+    if isinstance(obj, State):
+        return lambda part: part.name
+    if isinstance(obj, str):
+        return str.lower
+    return None
```

Comment:
I understand why you let this function return a callable now instead of using this function itself as the sorting key :) You want the sorting key to fall back to `None` ([the default](https://docs.python.org/3/library/functions.html#sorted)). It's a good idea, but I don't trust yet that it works as intended.
It seems to me that, in `_gen_sorting_key(group.states[i])`, `group.states[i]` is actually the list of items that is to be sorted, not the item itself.

So still, I would let this function return the value you want to sort itself (e.g. `name` for a `Particle` and `obj.lower()` for a string), so that you can just write `sorted(..., key=_gen_sortingkey)`. As fallback, return `obj` itself, as that would be the same behavior as `sorted(..., key=None)`. I also realize now (see https://github.com/ComPWA/qrules/pull/274#discussion_r1668674840) that this function does not need to have the responsibility to check whether the object actually supports comparison.